RFC: Default type alias type parameters (#61)
Roblox/luau: rfcs/syntax-default-type-alias-type-parameters.md

> # Default type alias type parameters

Reviewer: We'll probably want to apply the default type arguments to generic functions too. This title implies restricting this feature to type aliases only. I assume that's unintentional.

```suggestion
# Default type parameters
```

Also rename the file as well.

## Summary

Introduce syntax to provide default type values inside the type alias type parameter list.

## Motivation

Luau has support for type parameters for type aliases and functions.

> In languages with similar features like C++ and TypeScript, it is possible to specify default values and users with experience in those languages would like to use that in Luau.

Reviewer 2: ```suggestion
In languages with similar features like C++, Rust, flowtype, and TypeScript, it is possible to specify default values for looser coupling and easier composability, and users with experience in those languages would like to have these design capabilities Luau.
```

Author: I've included this, thank you.

## Design

If a default type argument value is assigned, following type parameters (on the right) must also have default type argument values.
```lua
type A<T, U = string, V> = ... -- not allowed
```

Default type argument values can reference type parameters which were defined earlier (to the left):
```lua
type A<T, U = T> = ...-- ok

type A<T, U = V, V = T> = ... -- not allowed
```

Default type arguments are not allowed for type pack parameters.

---

Syntax for type alias type argument is extended as follows:

```typearg ::= Name [`=' typeannotation]```

Instead of storing a simple array of names in AstStatTypeAlias, we will store an array of structs containing the name and an optional default type value.

When type alias is referenced, missing type parameters are replaced with default type values, if they are available.

If all type parameters have a default type value, it is now possible to reference that without providing a type parameter list:
```lua
type All<T = string, U = number> = ...

> local a: All -- ok

Reviewer 3: This would mean that there's no way to reference the actual generic type function, right? Is that valuable?

Author: If there a context in which you are able to do that? As far as I know you get an IncorrectGenericParameterCount error.

Reviewer 3: Makes sense.

local b: All<> -- not allowed
```

If type is exported from a module, default type parameter values will still be available when module is imported.

## Drawbacks

Previously, it was possible to reference type alias without type arguments inside that alias (recursive types).

> We disallow that now, but if we decide to return that, there will be an ambiguity if all type parameters have a default value.

Reviewer 3: Could we get an example of this?

Author: Wow, maybe this drawback is incorrect. I've rechecked change history by @asajeffrey and it doesn't look like this was ever possible (I though he fixed this).

I'll fix the text later, but here's an example of ambiguity in a different language:
```c++
template<typename T = int>
struct vector
{
    void foo()
    {
        vector v; // vector<T> instead of vector<int> even though generic type argument is not specified
    }
};
vector x; // vector<int>
```

Reviewer 3: Ah, that makes sense.

Author: I've changed the text and provided an example with theoretical Luau syntax.

## Alternatives

> We could use `:` instead of `=` to define default type parameter values.

Reviewer 3: Not fond of this. `:` implies the left hand side has a certain type. That doesn't make sense in the current type system for types (we don't have kinds). `=` seems sensible.

Reviewer: Yeah I'd prefer we use `=`. We might want to use `:` for bounded polymorphism.

Reviewer 2: I agree on both counts. Even if we don't like using `=` for defaulting, breathing room should be kept for expressing type constraints. It would be nice if this aligned with the syntax for default values of function arguments.

Reviewer 4: Seems worth adding some rationale to the RFC for why this route is not taken.

Author: I've removed `Alternatives` section and noted the decision between `:` and `=` in the `Design` section including a future path for bounded quantification.

Without default parameter values it's also harder to refactor the code as each type alias reference that uses 'common' type arguments has to be updated.

While previous example uses a concrete type for default type value, it should also be possible to reference generic types from the same list:
```lua
type Eq<T, U = T> = (l: T, r: U) -> boolean

local a: Eq<number> = ...
local b: Eq<number, string> = ...
```

Generic functions in Luau also have a type parameter list, but it's not possible to specify type arguments at the call site and because of that, default type parameter values for generic functions are not proposed.

## Design

If a default type parameter value is assigned, following type parameters (on the right) must also have default type parameter values.
```lua
type A<T, U = string, V> = ... -- not allowed
```

Default type parameter values can reference type parameters which were defined earlier (to the left):
```lua
type A<T, U = T> = ...-- ok

type A<T, U = V, V = T> = ... -- not allowed
```

Default type parameter values are also allowed for type packs:
```lua
type A<T, U... = ...string>        -- ok, variadic type pack

> type B<T, U... = (number, number)> -- ok, type pack with two elements

Reviewer 3: We don't have syntax for this at the moment. IMO this should be a separate proposal.

Author: Which part of the syntax exactly?

Reviewer 3: The `(T, T)` syntax.

Author: Fixed.
That's from return type annotation tail and is not accessible as a stand-alone syntax. I've removed this line.

type C<T, U... = string>           -- ok, type pack with one element

> type D<T..., U... = T>             -- ok

Reviewer 3: ```suggestion
type D<T..., U... = T...>          -- ok
```
We should be careful to disambiguate between `T` and `T...` (generic _type_ vs. generic _pack_)!

Author: Fixed.

If all type parameters have a default type value, it is now possible to reference that without providing a type parameter list:
```lua
type All<T = string, U = number> = ...

local a: All -- ok
local b: All<> -- not allowed
```

If type is exported from a module, default type parameter values will still be available when module is imported.

---
Type annotations in Luau are placed after `':'`, but we use `'='` here to assign a type value, not to imply that the type parameter on the left has a certain type.

Type annotation with `':'` could be used in the future for bounded quantification which is orthogonal to the default type value.

## Drawbacks

Other languages might allow references to the type alias without arguments inside the scope of that type alias to resolve into a recursive reference to the type alias with the same arguments.

While that is not allowed in Luau right now, if we decide to change that in the future, we will have an ambiguity when all type alias parameters have default values:
```lua
-- ok if we allow Type to mean Type<A, B>
type Type<A, B> = { x: number, b: Type? }

> -- ambiguity, Type could mean Type<number, string> or Type<A, B>

Reviewer 3: I think this is fine in terms of actual usability; you can just write `Type<A, B>` if you actually need the generic type, right?

Author: Yes.
